syntax = "proto3";

package nativedroid_server;
option java_package = "org.argus.nativedroid.server";

import "java_signatures.proto";

message GenSummaryRequest {
    string so_handle = 1;
    string jni_func = 2;
    jawa_core.MethodSignature method_signature = 3;
}

message GenSummaryResponse {
    string taint = 1;
    string summary = 2;
    int64 analyzed_instructions = 3;
}

message HasSymbolRequest {
    string so_handle = 1;
    string symbol = 2;
}

message HasSymbolResponse {
    bool has_symbol = 1;
}

message AnalyseNativeActivityRequest {
    string so_handle = 1;
    string custom_entry = 2;
}

message AnalyseNativeActivityResponse {
    int64 total_instructions = 1;
}

message LoadBinaryRequest {
    bytes buffer = 1;
}

message LoadBinaryResponse {
    string so_handle = 1;
    int32 length = 2;
}

service NativeDroidServer {
    rpc GenSummary(GenSummaryRequest) returns (GenSummaryResponse);
    rpc HasSymbol(HasSymbolRequest) returns (HasSymbolResponse);
    rpc AnalyseNativeActivity(AnalyseNativeActivityRequest) returns (AnalyseNativeActivityResponse);
    rpc LoadBinary(stream LoadBinaryRequest) returns (LoadBinaryResponse);
}